### PR TITLE
Imports reordering

### PR DESCRIPTION
**Pull request by anibyl**
Fix for the #471.

### CODE REVIEW

File: src/common/com/intellij/plugins/haxe/ide/HaxeImportOptimizer.java [hunk 1]
```diff
@@ -69,6 +82,49 @@ private static void optimizeImports(PsiFile file) {
...
+
+    for (HaxeImportStatementRegular currentImport : allImports) {
+      sortedImports.add(currentImport.getText());
+    }
+
+    Collections.sort(sortedImports);
+
+    final Document document = PsiDocumentManager.getInstance(file.getProject()).getDocument(file);
+    if (document != null) {
+      final PsiDocumentManager documentManager = PsiDocumentManager.getInstance(file.getProject());
```
Comment (EBatTiVo) [inline src/common/com/intellij/plugins/haxe/ide/HaxeImportOptimizer.java]: These three lines will read a lot better if you did (paraphrased):

docMgr = getInstance(file.getProject);
document = docMgr.getDocument(file);
if (document) { ...

**Comment by EBatTiVo** [comment]
There is a fundamental flaw with re-ordering: it assumes the common practice of placing all imports in a single block at the top of a file.  This is not necessarily the case, and we should detect whether any code beside import statements and blank lines (and maybe comments, see below) exists between import statements, and refuse to re-order them, if so.  We should not be able to delete code using this function.

I am wondering if we can't do a better job with remove unused and reordering with regard to comments.  That is, when a comment immediately precedes an import, should we try to keep the comment along with it.  As it is, this code just deletes comments (or any other code) that appears within the import "block."  (There is no real block, it's just a conceptual idea.)  We really should either attempt to keep comments with the import, or block the re-ordering if comments appear.

**Comment by anibyl** [comment]
You are right, it removes any text (including code) between the first and the last import. I did it just for smaller execution time, I didn't even think that somebody place a code between them :)
So, we can delete every import personally and the place them together at the top of file.

Keeping the comments? Hmm, interesting. I can investigate if it is possible to do it without big additional executing time.

What else should be kept?

Also, if we have such file, do we need to reorder imports only in these divided "blocks":
```
import a;
import b;
class A {}

import c;
import d;
class B {}
```
? Or just put them together like
```
import a;
import b;
import c;
import d;

class A {}

class B {}
```

PS: I am worried about execution time because removing unused imports in the file i tested on takes more than 5s and makes the IDEA hang.

**Comment by anibyl** [comment]
I took a look at the java imports optimization:
```
// First comment
import android.view.Menu;
// Second comment
import android.content.Intent;
import android.net.Uri;
import android.view.MenuItem;
import android.view.View;
// Third comment
import android.widget.Toolbar;
import android.os.Build;
import android.os.Bundle;
import android.app.ActionBar;
// Fourth comment
import android.support.annotation.NonNull;
import android.text.TextUtils;
// Fifth comment
```
became
```
// First comment

import android.app.ActionBar;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toolbar;

// Second comment
// Third comment
// Fourth comment
// Fifth comment
```

Are you sure we should care about comments?

**Comment by EBatTiVo** [comment]
My feeling is that the Java implementation isn't really great, anyway.  We can do better.  It's just a cost/benefit tradeoff.  I frequently write comments around imports, particularly when the import file name isn't clear about what it contains.  In those cases, I'd prefer to keep the comment with the import.  You might also want to check (or not, I guess :P) whether there are any language style settings having to do with sorting imports, such as adding newlines between unrelated groups.  If there are, it might be nice to obey them.  Other than the possibility of deleting code, this is still a nice feature just as it is, so I don't want to let the perfect be the enemy of the good.

I would recommend that we simply do not do the re-ordering in the case where code appears between them.  Folks who write code that intersperses code with imports are generally looking to create, override, or affect specific behaviors that could (would!) be changed by re-ordering.  You can simply look for non-whitespace elements in the PSI between the first and last import elements.

Also, if things are taking 5s per file, I would take a look at the scanning algorithm.  Is it blocking waiting for a write lock on the UI thread when it only needs read access?  We only need write access to the file when we actually want to change something.  At any rate, if the process does take too long, we can run it in a progress dialog so the user doesn't think IDEA has hung.

Another thought, perhaps every file change is sending messages to the annotator, etc. to update the screen between each add?  If that turns out to be the case, then you can build the entire new block as a single string and add it in a single PSI insertion operation at the end.  I guess we'll have to profile the code to get a really good view of what is happening.

**Comment by anibyl** [comment]
Alright, I'll add a check for different PSI elements in imports “block”. It will not rearrange imports if there is something unusual, otherwise I need to check all the possible cases and I am sure that I can miss something :)

I took a look at imports settings of supported languages — they are completely different. We can create “Imports” page in Haxe code style and put necessary options there (such as “Enable reordering” :D). Blank lines between import groups should be declared in that place also.

About hanging: unused imports removing does many resolvings in case your class is huge (`HaxeImportUtil.findUnusedImports(file)`, `HaxeImportUtil.findUnusedInImports(file)` and `HaxeImportUtil.findUnusedInImportsWithWildcards(file)`). It was already there before this PR, I just added a TODO “optimize me”.
IDEA hangs in debug mode, probably it is better in release.

**Comment by anibyl** [comment]
@EBatTiVo, please review PR again.

**Comment by anibyl** [comment]
Up.

**Pull request merged**